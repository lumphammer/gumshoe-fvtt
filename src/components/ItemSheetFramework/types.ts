export enum ItemSheetMode {
  Main = "main",
  Config = "config",
}
