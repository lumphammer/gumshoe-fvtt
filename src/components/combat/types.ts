export type TurnInfo = {
  id: string;
  name: string;
  img: string;
  active: boolean;
  defeated: boolean;
  hidden: boolean;
  initiative: number | undefined | null;
  resource: number | boolean | string | null;
  effects: foundry.documents.ActiveEffect[];
  passingTurnsRemaining: number;
  totalPassingTurns: number;
};
