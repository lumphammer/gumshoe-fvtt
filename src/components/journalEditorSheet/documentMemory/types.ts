import { Change } from "textdiff-create";

/**
 * An edit to a document. The change itself is a diff from textdiff-create, and
 * we also store the timestamp of the edit.
 */
export interface Edit {
  /** The actual change, as a diff */
  changes: Change[];
  /** The timestamp of the edit, in epoch seconds */
  timestamp: number;
  /** The serial number of the edit, unique for the document */
  serial: number;
}

/**
 * A stack of edits in a document memory. Each stack represents an older stage
 * of the memory and has pointer the next, even, older, one. So, the freshest
 * edits are in the first stack. Every time that fills once, a  combnined edit
 * isadded to the next stack, recursively.
 */
export interface Stack {
  /** The edits in this stack. Will always have length == period of the parent
   * memory
   */
  edits: Array<Edit>;
  /**
   * The state of the document the last time this stack pushed down to the
   * next stack.
   */
  bombBay: Array<Edit>;
  /**
   * A pointer to the next stack, to which we will dump edits every time we have
   * recorded `period` edits.
   */
  next: Stack | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 * Any commit can be restored by walking the stacks from the start.
 */
export interface BareDocumentMemory {
  /**
   * A pointer to the first stack in the memory, the one which will contain the
   * most recent edits.
   */
  stack: Stack;
  /**
   * The curent serial number.
   */
  serial: number;
  /**
   * The period of this memory, i.e. how many edits are stored in each stack.
   */
  period: number;
  /**
   * Limit on maximum depth of stacks. The total number of edits stored will be
   * period * maxDepth, representing
   * period^maxDepth + period^(maxDepth - 1) + ... + period^1
   * total edits.
   */
  maxDepth: number | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 * Contains snapshots of each stack to accelerate pushes and restores.
 */
export interface DocumentMemory extends BareDocumentMemory {
  /**
   * a cache of snapshots of each level of the document
   */
  snapshots: string[];
  /**
   * A snapshot of the most recent state of the document
   */
  state: string;
}

/**
 * This type is used in to store the document memories globally. The key is the
 * id of the document (aka JournalEntryPage) plus the id of the and the
 */
export type DocumentMemoryCollection = Record<string, BareDocumentMemory>;
