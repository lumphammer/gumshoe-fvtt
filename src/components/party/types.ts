import { AbilityDataSource, AbilityType } from "../../types";

export const typeHeaderKey = "typeHeader" as const;
export const categoryHeaderKey = "categoryHeader" as const;
export const abilityRowkey = "abilityRowString" as const;
export type TypeHeaderData = {
  rowType: typeof typeHeaderKey;
  abilityType: AbilityType;
};
export type CategoryHeaderData = {
  rowType: typeof categoryHeaderKey;
  category: string;
};
export type ActorAbilityInfo = {
  actorId: string;
  abilityId: string | undefined;
  rating: number | undefined;
};
export type AbilityRowData = {
  rowType: typeof abilityRowkey;
  abilityDataSource: AbilityDataSource;
  actorInfo: {
    [actorId: string]: ActorAbilityInfo;
  };
  total: number;
};
export type RowData = TypeHeaderData | CategoryHeaderData | AbilityRowData;
export const isTypeHeader = (data: RowData): data is TypeHeaderData =>
  data.rowType === typeHeaderKey;
export const isCategoryHeader = (data: RowData): data is CategoryHeaderData =>
  data.rowType === categoryHeaderKey;
