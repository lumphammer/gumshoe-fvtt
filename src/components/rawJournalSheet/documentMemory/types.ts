import { Change } from "textdiff-create";

/**
 * An edit to a document. The change itself is a diff from textdiff-create, and
 * we also store the timestamp of the edit.
 */
export interface Edit {
  /** The actual change, as a diff */
  changes: Change[];
  /** The timestamp of the edit, in epoch seconds */
  timestamp: number;
}

/**
 * A stack of edits in a document memory. Each stack represents an older stage
 * of the memory and has pointer the next, even, older, one. So, the freshest
 * edits are in the first stack. Every time that fills once, a  combnined edit
 * isadded to the next stack, recursively.
 */
export interface Stack {
  /** The edits in this stack. Will always have length == period of the parent
   * memory
   */
  edits: Array<Edit | null>;
  /**
   * The state of the document the last time this stack was pushed to. We store
   * this so that we can calculate a diff when a new state comes in. The
   * alternative being to store reversible diffs, i.e. diffs that contain the
   * actual text from deletions so they can be reversed, which could make diffs
   * very large if there are lots of large changes.
   * XXX can we do anything cleverer?
   */
  // lastPushSnapshot: string;
  /**
   * The state of the document the last time this stack pushed down to the
   * next stack.
   */
  bombBay: Change[];
  /**
   * A pointer to the next stack, to which we will dump edits every time we have
   * recorded `period` edits.
   */
  next: Stack | null;
}

/**
 * A document's edit memory. Older edits are stored with less granularity.
 */
export interface DocumentMemory {
  /**
   * A pointer to the first stack in the memory, the one which will contain the
   * most recent edits.
   */
  stack: Stack;
  /**
   * The curent serial number.
   */
  serial: number;
  /**
   * The period of this memory, i.e. how many edits are stored in each stack.
   */
  period: number;
  /**
   * Limit on maximum depth of stacks. The total number of edits stored will be
   * period * maxDepth, representing
   * period^maxDepth + period^(maxDepth - 1) + ... + period^1
   * total edits.
   */
  maxDepth: number | null;
  /**
   * A snapshot of the most recent state of the document
   */
  state: string;
  /**
   * a cache of snapshots of each level of the document
   */
  snapshots: string[];
}
