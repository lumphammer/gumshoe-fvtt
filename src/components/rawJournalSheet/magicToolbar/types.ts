import { ReactNode } from "react";

export type MagicToolbarContentEntry = {
  category: string;
  content: ReactNode;
};

export type MagicToolbarContent = Record<string, MagicToolbarContentEntry>;

export type MagicToolBarRegister = (
  id: string,
  content: MagicToolbarContentEntry,
) => void;

export type MagicToolBarUnregister = (id: string) => void;

export type MagicToolbarRegisterContextDataValue = {
  register: MagicToolBarRegister;
  unregister: MagicToolBarUnregister;
};
