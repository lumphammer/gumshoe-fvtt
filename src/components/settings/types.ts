import { SettingsDict } from "../../settings";

export type Setters = {
  [k in keyof SettingsDict]: (newVal: SettingsDict[k]) => void;
};

export interface State {
  settings: SettingsDict;
}

export type PcOrNpc = "pcStats" | "npcStats";
