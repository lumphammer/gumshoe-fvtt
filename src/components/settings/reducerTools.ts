import { EquipmentFieldMetadata } from "@lumphammer/investigator-fvtt-types";
import { Draft, produce } from "immer";

/**
 * A minimal case for all actions - there will always be a `type` and a
 * `payload` property, but the payload may be `undefined`.
 */
export type AnyAction = {
  type: string;
  payload?: unknown;
};

/**
 * A minimal reimplementation of the `createAction` function from
 * `@reduxjs/toolkit`.
 *
 * Given a unique string (danger will robinson!) and a reducer function, returns
 * an object with
 *
 *  * a `create` method that returns an action object
 *  * an `apply` function, which take a state and an action and applies the
 *    relevant reducer if the action matches the case
 */
function createCase<S, P = void>(
  type: string,
  reducer: (draft: S, payload: P) => void,
) {
  const create = (payload: P) => ({ type, payload });
  const match = (action: AnyAction): action is { type: string; payload: P } =>
    action.type === type;
  const apply = (state: S, action: AnyAction) => {
    if (match(action)) {
      return reducer ? reducer(state, action.payload) : state;
    }
    return state;
  };
  return {
    /** Create an action (to be dispatched with `dispatch` from `useReducer` */
    create,
    /**
     * given a mutable draft object of S, and action, mutate the draft if the
     * action is from this slice.
     */
    apply,
  };
}

type Reducers<S> = { [key: string]: (state: Draft<S>, payload?: any) => void };

export type CreateSliceArgs = {
  onError?: (error: unknown) => void;
  log?: (...args: unknown[]) => void;
};

/**
 * A minimal reimagination of the `createSlice` function from
 * `@reduxjs/toolkit`.
 * This function is curried so that we can specify S (state) but allow R
 * (reducers) to be inferred.
 */
export const createSlice =
  <S extends object>(args: CreateSliceArgs = {}) =>
  <R extends Reducers<S>>(reducers: R) => {
    // turn all the reducers into cases (so they can be created and applied)
    const sliceCases = Object.entries(reducers).map(([key, reducer]) => {
      const action = createCase(key, reducer);
      return [key, action] as const;
    });
    // turn those cases back into an object of creators.
    const creators = Object.fromEntries(
      sliceCases.map(([key, action]) => [key, action.create]),
    ) as {
      // we need this mad cast because `map` can't give us accurate types.
      // we're just knitting together the `typeof createCase<...>["create"]`
      // and working out the payload type using inference.
      [key in keyof R]: ReturnType<
        typeof createCase<
          S,
          R[key] extends (state: infer S1) => void
            ? void
            : R[key] extends (state: infer S1, payload: infer P1) => void
              ? P1
              : never
        >
      >["create"];
    };
    // `useDispatch` takes the initial state as an argument, so we don't need to
    // provide a default argument for `state` here (unlike e.g. a Redux
    // reducer).
    const reducer = (state: S, action: AnyAction) => {
      try {
        const newState = produce(state, (draft) => {
          for (const [, sliceCase] of sliceCases) {
            sliceCase.apply(draft, action);
          }
        });
        if (args.log) {
          args.log(action, state, newState);
        }
        return newState;
      } catch (e) {
        // in the event of an error, we will try to keep going rather than just
        // exploding
        if (args.onError) {
          args.onError(e);
        }
        return state;
      }
    };

    return {
      /**
       * A dictionary of action creators, keyed by the name of the case.
       */
      creators,
      /**
       * The reducer function for this slice.
       */
      reducer,
    };
  };

export function assertNumericFieldOkayness(
  field: EquipmentFieldMetadata | undefined,
  id: string,
  value: unknown | undefined,
): asserts field is Extract<EquipmentFieldMetadata, { type: "number" }> {
  if (field === undefined) {
    throw new Error(`No field with id ${id}`);
  }
  if (field.type !== "number") {
    throw new Error(`Cannot set min/max on field type ${field.type}`);
  }
  if (typeof value !== "number" && value !== undefined) {
    throw new Error(
      `Invalid value ${value} for field ${field.name} (must be a number)`,
    );
  }
}
