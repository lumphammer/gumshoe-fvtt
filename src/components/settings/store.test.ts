import { diff } from "jest-diff";
import { afterAll, beforeEach, describe, expect, it, vi } from "vitest";

import { pathOfCthulhuPreset } from "../../presets";
import { AnyAction } from "./reducerTools";
import { createSystemSlice } from "./store";
import { State } from "./types";

type TestTuple = [string, AnyAction];

const initialState: State = {
  settings: {
    pcStats: {
      pcStat0: {
        min: 0,
        max: 10,
        default: 3,
        name: "PC Statisitic Zero",
      },
    },
    npcStats: {
      npcStat0: {
        min: 0,
        max: 10,
        default: 3,
        name: "NPC Statisitic Zero",
      },
    },
    systemPreset: "dnd5e",
    combatAbilities: [],
    customThemePath: "",
    equipmentCategories: {
      category0: {
        name: "initial category",
        fields: {
          field0: {
            name: "Initial field",
            type: "string",
            default: "default value",
          },
          field1: {
            name: "Initial field",
            type: "number",
            min: 0,
            max: 10,
            default: 1,
          },
        },
      },
      category1: {
        name: "initial category",
        fields: {},
      },
    },
    debugTranslations: false,
    useBoost: false,
    defaultThemeName: "",
    useMwInjuryStatus: false,
    useMwStyleAbilities: false,
    useNpcCombatBonuses: false,
    generalAbilityCategories: [],
    migrationFlags: {
      actor: {},
      item: {},
      compendium: {},
      journal: {},
      macro: {},
      scene: {},
      rollTable: {},
      playlist: {},
      world: {},
    },
    useTurnPassingInitiative: false,
    genericOccupation: "",
    investigativeAbilityCategories: [],
    longNotes: [],
    mwHiddenShortNotes: [],
    mwUseAlternativeItemTypes: false,
    newNPCPacks: [],
    newPCPacks: [],
    occupationLabel: "",
    personalDetails: [],
    showEmptyInvestigativeCategories: false,
    systemMigrationVersion: "",
    abilityCategories: "",
    shortNotes: [],
    firstRun: true,
    journalMemories: {},
  },
};

// const initialStateJSON = JSON.stringify(initialState, null, 2);

// no-op used later to prevent jest-diff trying to color output
const noColor = (x: string) => x;

// options for jest-diff
const diffOptions = {
  // don't colourise the diff
  aColor: noColor,
  bColor: noColor,
  changeColor: noColor,
  commonColor: noColor,
  patchColor: noColor,
  // show 3 lines of context around each change
  contextLines: 3,
  // don't show the entire object in the diff
  expand: false,
  omitAnnotationLines: true,
  // keep object keys in the original order (so key order changes will be
  // detected)
  compareKeys: () => 0,
};

// mock the reducer callbacks
const logFn = vi.fn();
const onErrorFn = vi.fn();

const slice = createSystemSlice({
  log: logFn,
  onError: onErrorFn,
});

beforeEach(() => {
  logFn.mockReset();
  onErrorFn.mockReset();
});

// we need to mock nanoid so that we can predictably generate ids
// I tried putting this in beforeAll but for some reason we weren't seeing the
// mock in the test
vi.mock("nanoid", () => {
  let count = 0;
  return {
    __esModule: true,
    nanoid: () => "id" + count++,
  };
});

afterAll(() => {
  vi.resetModules();
});

/*
 * DIFF SNAPSHOT TESTS!
 *
 * Rather than maintain a manual list of expected values,
 * we can just snapshot the diff between the initial state and the state after
 * the action is applied. This means the tests are not super brittle in the
 * event of the structure changing, but it does require you to manually check
 * the diff to make sure it's what you expect.
 *
 * The snapshots are stored in the __snapshots__ folder.
 *
 * To update the snapshots, run `npx jest -u`
 *
 * AND PLEASE CHECK THE DIFFS AFTER UPDATING! Your git tool of choice will help
 * here, so you can just compare the diffs that have changed. Yes, it might feel
 * a bit weird looking at "diffs of diffs", but you'll get the hang of it.
 */
describe("reducer", () => {
  it("should ignore unknown actions", () => {
    const result = slice.reducer(initialState, { type: "FOOEY" });
    // no point snapshotting this
    expect(result).toEqual(initialState);
  });

  it.each<TestTuple>([
    [
      "set a value using setSome",
      slice.creators.setSome({ newSettings: { genericOccupation: "foo" } }),
    ],
    ["add a category", slice.creators.addCategory()],
    ["delete a category", slice.creators.deleteCategory({ id: "category0" })],
    [
      "rename a category",
      slice.creators.renameCategory({ id: "category0", newName: "foo" }),
    ],
    [
      "change a category id",
      slice.creators.changeCategoryId({
        oldCategoryId: "category0",
        newCategoryId: "foo",
      }),
    ],
    [
      "move a category up",
      slice.creators.moveCategoryUp({ categoryId: "category1" }),
    ],
    [
      "move a category down",
      slice.creators.moveCategoryDown({ categoryId: "category0" }),
    ],
    ["add a field", slice.creators.addField({ categoryId: "category0" })],
    [
      "delete a field",
      slice.creators.deleteField({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "rename a field",
      slice.creators.renameField({
        categoryId: "category0",
        fieldId: "field0",
        newName: "foo",
      }),
    ],
    [
      "change a field id",
      slice.creators.changeFieldId({
        categoryId: "category0",
        fieldId: "field0",
        newFieldId: "foo",
      }),
    ],
    [
      "set a field type",
      slice.creators.setFieldType({
        categoryId: "category0",
        fieldId: "field0",
        newType: "number",
      }),
    ],
    [
      "set a field default",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field0",
        newDefault: "foo",
      }),
    ],
    [
      "set a field minimum",
      slice.creators.setFieldMin({
        categoryId: "category0",
        fieldId: "field1",
        newMin: 2,
      }),
    ],
    [
      "set a field maximum",
      slice.creators.setFieldMax({
        categoryId: "category0",
        fieldId: "field1",
        newMax: 2,
      }),
    ],
    [
      "apply a preset",
      slice.creators.applyPreset({
        preset: pathOfCthulhuPreset,
        presetId: "pathOfCthulhu",
      }),
    ],
    [
      "add a stat",
      slice.creators.addStat({
        which: "pcStats",
      }),
    ],
    [
      "set a stat minimum",
      slice.creators.setStatMin({
        which: "pcStats",
        statId: "pcStat0",
        newMin: 1,
      }),
    ],
    [
      "set a stat maximum",
      slice.creators.setStatMax({
        which: "pcStats",
        statId: "pcStat0",
        newMax: 1,
      }),
    ],
    [
      "move a field up",
      slice.creators.moveFieldUp({
        categoryId: "category0",
        fieldId: "field1",
      }),
    ],
    [
      "move a field down",
      slice.creators.moveFieldDown({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "set a stat default",
      slice.creators.setStatDefault({
        which: "pcStats",
        statId: "pcStat0",
        newDefault: 1,
      }),
    ],
    [
      "set a stat name",
      slice.creators.setStatName({
        which: "pcStats",
        statId: "pcStat0",
        newName: "foo",
      }),
    ],
    [
      "delete a stat",
      slice.creators.deleteStat({
        which: "pcStats",
        statId: "pcStat0",
      }),
    ],
    [
      "set a stat id",
      slice.creators.setStatId({
        which: "pcStats",
        oldStatId: "pcStat0",
        newStatId: "foo",
      }),
    ],
  ])("should %s", (name, action) => {
    // * why snapshots of diffs?
    //   * they're smaller and easier to read than snapshotting the entire
    //     output object. You can peruse the snapshots file and see the
    //     *changes* caused by the action, rather than the entire state.
    // * why a patch formatted diff?
    //   * first attempt used an object diff (from `just-diff`) but that does
    //     not detect changes in key order.
    const result = slice.reducer(initialState, action);
    const diffs = diff(initialState, result, diffOptions);
    expect(diffs).toMatchSnapshot();
    expect(onErrorFn).not.toHaveBeenCalled();
    expect(logFn).toHaveBeenCalledTimes(1);
    expect(logFn).toHaveBeenCalledWith(action, initialState, result);
  });

  // error cases
  it.each<TestTuple>([
    [
      "moving a category up when it's already at the top",
      slice.creators.moveCategoryUp({ categoryId: "category0" }),
    ],
    [
      "moving a category down when it's already at the bottom",
      slice.creators.moveCategoryDown({ categoryId: "category1" }),
    ],
    [
      "moving a field up when it's already at the top",
      slice.creators.moveFieldUp({
        categoryId: "category0",
        fieldId: "field0",
      }),
    ],
    [
      "moving a field down when it's already at the bottom",
      slice.creators.moveFieldDown({
        categoryId: "category0",
        fieldId: "field1",
      }),
    ],
    [
      "setting a field minimum to a value greater than its maximum",
      slice.creators.setFieldMin({
        categoryId: "category0",
        fieldId: "field1",
        newMin: 20,
      }),
    ],
    [
      "setting a stat minimum to a value greater than its maximum",
      slice.creators.setStatMin({
        which: "pcStats",
        statId: "pcStat0",
        newMin: 20,
      }),
    ],
    [
      "setting a stat maximum to a value less than its minimum",
      slice.creators.setStatMax({
        which: "pcStats",
        statId: "pcStat0",
        newMax: -10,
      }),
    ],
    [
      "changing a catrgory id to one that already exists",
      slice.creators.changeCategoryId({
        oldCategoryId: "category0",
        newCategoryId: "category1",
      }),
    ],
    [
      "changing a field id to one that already exists",
      slice.creators.changeFieldId({
        categoryId: "category0",
        fieldId: "field0",
        newFieldId: "field1",
      }),
    ],
    [
      "setting a numeric default for a string field",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field0",
        newDefault: 1,
      }),
    ],
    [
      "setting a boolean default for a string field",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field0",
        newDefault: true,
      }),
    ],
    [
      "setting a string default for a numeric field",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field1",
        newDefault: "foo",
      }),
    ],
    [
      "setting a boolean default for a numeric field",
      slice.creators.setFieldDefault({
        categoryId: "category0",
        fieldId: "field1",
        newDefault: true,
      }),
    ],
    [
      "deliberately throwing an error",
      slice.creators.throwError({
        message: "foo",
      }),
    ],
  ])("should error when %s", (name, action) => {
    const result = slice.reducer(initialState, action);
    expect(result).toEqual(initialState);
    expect(logFn).not.toHaveBeenCalled();
    expect(onErrorFn).toHaveBeenCalledTimes(1);
  });
});
