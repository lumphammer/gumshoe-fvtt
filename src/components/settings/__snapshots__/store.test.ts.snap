// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`reducer > should add a category 1`] = `
"@@ -41,6 +41,10 @@
          "name": "initial category",
          "fields": Object {},
        },
+       "id0": Object {
+         "name": "New category",
+         "fields": Object {},
+       },
      },
      "debugTranslations": false,
      "useBoost": false,"
`;

exports[`reducer > should add a field 1`] = `
"@@ -35,6 +35,11 @@
              "max": 10,
              "default": 1,
            },
+           "id1": Object {
+             "name": "New Field",
+             "type": "string",
+             "default": "",
+           },
          },
        },
        "category1": Object {"
`;

exports[`reducer > should add a stat 1`] = `
"@@ -7,6 +7,10 @@
          "default": 3,
          "name": "PC Statisitic Zero",
        },
+       "stat1": Object {
+         "name": "",
+         "default": 0,
+       },
      },
      "npcStats": Object {
        "npcStat0": Object {"
`;

exports[`reducer > should apply a preset 1`] = `
"@@ -1,54 +1,56 @@
  Object {
    "settings": Object {
      "pcStats": Object {
-       "pcStat0": Object {
-         "min": 0,
-         "max": 10,
+       "hitThreshold": Object {
+         "name": "Hit threshold",
          "default": 3,
-         "name": "PC Statisitic Zero",
        },
      },
      "npcStats": Object {
-       "npcStat0": Object {
-         "min": 0,
-         "max": 10,
+       "hitThreshold": Object {
+         "name": "Hit threshold",
          "default": 3,
-         "name": "NPC Statisitic Zero",
        },
+       "armor": Object {
+         "name": "Armor",
+         "default": 0,
        },
-     "systemPreset": "dnd5e",
-     "combatAbilities": Array [],
-     "customThemePath": "",
-     "equipmentCategories": Object {
-       "category0": Object {
-         "name": "initial category",
-         "fields": Object {
-           "field0": Object {
-             "name": "Initial field",
-             "type": "string",
-             "default": "default value",
+       "alertness": Object {
+         "name": "Alertness",
+         "default": 0,
        },
-           "field1": Object {
-             "name": "Initial field",
-             "type": "number",
-             "min": 0,
-             "max": 10,
-             "default": 1,
+       "stealth": Object {
+         "name": "Stealth",
+         "default": 0,
        },
+       "stabilityLoss": Object {
+         "name": "Stability Loss",
+         "default": 0,
        },
      },
-       "category1": Object {
-         "name": "initial category",
+     "systemPreset": "pathOfCthulhu",
+     "combatAbilities": Array [
+       "Scuffling",
+       "Weapons",
+       "Firearms",
+       "Athletics",
+     ],
+     "customThemePath": "",
+     "equipmentCategories": Object {
+       "general": Object {
+         "name": "General",
          "fields": Object {},
        },
      },
      "debugTranslations": false,
      "useBoost": false,
-     "defaultThemeName": "",
+     "defaultThemeName": "tealTheme",
      "useMwInjuryStatus": false,
      "useMwStyleAbilities": false,
      "useNpcCombatBonuses": false,
-     "generalAbilityCategories": Array [],
+     "generalAbilityCategories": Array [
+       "General",
+     ],
      "migrationFlags": Object {
        "actor": Object {},
        "item": Object {},
@@ -61,20 +63,40 @@
        "world": Object {},
      },
      "useTurnPassingInitiative": false,
-     "genericOccupation": "",
-     "investigativeAbilityCategories": Array [],
-     "longNotes": Array [],
+     "genericOccupation": "Investigator",
+     "investigativeAbilityCategories": Array [
+       "Academic",
+       "Interpersonal",
+       "Technical",
+     ],
+     "longNotes": Array [
+       "Notes, Contacts etc.",
+       "Occupational Benefits",
+       "Pillars of Sanity",
+       "Sources of Stability",
+     ],
      "mwHiddenShortNotes": Array [],
      "mwUseAlternativeItemTypes": false,
-     "newNPCPacks": Array [],
-     "newPCPacks": Array [],
-     "occupationLabel": "",
-     "personalDetails": Array [],
+     "newNPCPacks": Array [
+       "investigator.opponentAbilities",
+     ],
+     "newPCPacks": Array [
+       "investigator.pathOfCthulhuAbilities",
+     ],
+     "occupationLabel": "Occupation",
+     "personalDetails": Array [
+       Object {
+         "name": "Drive",
+         "type": "item",
+       },
+     ],
      "showEmptyInvestigativeCategories": false,
      "systemMigrationVersion": "",
      "abilityCategories": "",
      "shortNotes": Array [],
      "firstRun": true,
      "journalMemories": Object {},
+     "schemaVersion": "v1",
+     "displayName": "Trail of Cthulhu (built-in)",
    },
  }"
`;

exports[`reducer > should change a category id 1`] = `
"@@ -20,7 +20,7 @@
      "combatAbilities": Array [],
      "customThemePath": "",
      "equipmentCategories": Object {
-       "category0": Object {
+       "foo": Object {
          "name": "initial category",
          "fields": Object {
            "field0": Object {"
`;

exports[`reducer > should change a field id 1`] = `
"@@ -23,12 +23,12 @@
        "category0": Object {
          "name": "initial category",
          "fields": Object {
-           "field0": Object {
+           "foo": Object {
              "name": "Initial field",
              "type": "string",
              "default": "default value",
            },
-           "field1": Object {
+           "field0": Object {
              "name": "Initial field",
              "type": "number",
              "min": 0,"
`;

exports[`reducer > should delete a category 1`] = `
"@@ -20,23 +20,6 @@
      "combatAbilities": Array [],
      "customThemePath": "",
      "equipmentCategories": Object {
-       "category0": Object {
-         "name": "initial category",
-         "fields": Object {
-           "field0": Object {
-             "name": "Initial field",
-             "type": "string",
-             "default": "default value",
-           },
-           "field1": Object {
-             "name": "Initial field",
-             "type": "number",
-             "min": 0,
-             "max": 10,
-             "default": 1,
-           },
-         },
-       },
        "category1": Object {
          "name": "initial category",
          "fields": Object {},"
`;

exports[`reducer > should delete a field 1`] = `
"@@ -23,11 +23,6 @@
        "category0": Object {
          "name": "initial category",
          "fields": Object {
-           "field0": Object {
-             "name": "Initial field",
-             "type": "string",
-             "default": "default value",
-           },
            "field1": Object {
              "name": "Initial field",
              "type": "number","
`;

exports[`reducer > should delete a stat 1`] = `
"@@ -1,13 +1,6 @@
  Object {
    "settings": Object {
-     "pcStats": Object {
-       "pcStat0": Object {
-         "min": 0,
-         "max": 10,
-         "default": 3,
-         "name": "PC Statisitic Zero",
-       },
-     },
+     "pcStats": Object {},
      "npcStats": Object {
        "npcStat0": Object {
          "min": 0,"
`;

exports[`reducer > should move a category down 1`] = `
"@@ -20,6 +20,10 @@
      "combatAbilities": Array [],
      "customThemePath": "",
      "equipmentCategories": Object {
+       "category1": Object {
+         "name": "initial category",
+         "fields": Object {},
+       },
        "category0": Object {
          "name": "initial category",
          "fields": Object {
@@ -34,12 +38,8 @@
              "min": 0,
              "max": 10,
              "default": 1,
-           },
            },
          },
-       "category1": Object {
-         "name": "initial category",
-         "fields": Object {},
        },
      },
      "debugTranslations": false,"
`;

exports[`reducer > should move a category up 1`] = `
"@@ -20,6 +20,10 @@
      "combatAbilities": Array [],
      "customThemePath": "",
      "equipmentCategories": Object {
+       "category1": Object {
+         "name": "initial category",
+         "fields": Object {},
+       },
        "category0": Object {
          "name": "initial category",
          "fields": Object {
@@ -34,12 +38,8 @@
              "min": 0,
              "max": 10,
              "default": 1,
-           },
            },
          },
-       "category1": Object {
-         "name": "initial category",
-         "fields": Object {},
        },
      },
      "debugTranslations": false,"
`;

exports[`reducer > should move a field down 1`] = `
"@@ -23,17 +23,17 @@
        "category0": Object {
          "name": "initial category",
          "fields": Object {
-           "field0": Object {
-             "name": "Initial field",
-             "type": "string",
-             "default": "default value",
-           },
            "field1": Object {
              "name": "Initial field",
              "type": "number",
              "min": 0,
              "max": 10,
              "default": 1,
+           },
+           "field0": Object {
+             "name": "Initial field",
+             "type": "string",
+             "default": "default value",
            },
          },
        },"
`;

exports[`reducer > should move a field up 1`] = `
"@@ -23,17 +23,17 @@
        "category0": Object {
          "name": "initial category",
          "fields": Object {
-           "field0": Object {
-             "name": "Initial field",
-             "type": "string",
-             "default": "default value",
-           },
            "field1": Object {
              "name": "Initial field",
              "type": "number",
              "min": 0,
              "max": 10,
              "default": 1,
+           },
+           "field0": Object {
+             "name": "Initial field",
+             "type": "string",
+             "default": "default value",
            },
          },
        },"
`;

exports[`reducer > should rename a category 1`] = `
"@@ -21,7 +21,7 @@
      "customThemePath": "",
      "equipmentCategories": Object {
        "category0": Object {
-         "name": "initial category",
+         "name": "foo",
          "fields": Object {
            "field0": Object {
              "name": "Initial field","
`;

exports[`reducer > should rename a field 1`] = `
"@@ -24,7 +24,7 @@
          "name": "initial category",
          "fields": Object {
            "field0": Object {
-             "name": "Initial field",
+             "name": "foo",
              "type": "string",
              "default": "default value",
            },"
`;

exports[`reducer > should set a field default 1`] = `
"@@ -26,7 +26,7 @@
            "field0": Object {
              "name": "Initial field",
              "type": "string",
-             "default": "default value",
+             "default": "foo",
            },
            "field1": Object {
              "name": "Initial field","
`;

exports[`reducer > should set a field maximum 1`] = `
"@@ -32,7 +32,7 @@
              "name": "Initial field",
              "type": "number",
              "min": 0,
-             "max": 10,
+             "max": 2,
              "default": 1,
            },
          },"
`;

exports[`reducer > should set a field minimum 1`] = `
"@@ -31,7 +31,7 @@
            "field1": Object {
              "name": "Initial field",
              "type": "number",
-             "min": 0,
+             "min": 2,
              "max": 10,
              "default": 1,
            },"
`;

exports[`reducer > should set a field type 1`] = `
"@@ -25,8 +25,8 @@
          "fields": Object {
            "field0": Object {
              "name": "Initial field",
-             "type": "string",
-             "default": "default value",
+             "type": "number",
+             "default": 0,
            },
            "field1": Object {
              "name": "Initial field","
`;

exports[`reducer > should set a stat default 1`] = `
"@@ -4,7 +4,7 @@
        "pcStat0": Object {
          "min": 0,
          "max": 10,
-         "default": 3,
+         "default": 1,
          "name": "PC Statisitic Zero",
        },
      },"
`;

exports[`reducer > should set a stat id 1`] = `
"@@ -1,7 +1,7 @@
  Object {
    "settings": Object {
      "pcStats": Object {
-       "pcStat0": Object {
+       "foo": Object {
          "min": 0,
          "max": 10,
          "default": 3,"
`;

exports[`reducer > should set a stat maximum 1`] = `
"@@ -3,7 +3,7 @@
      "pcStats": Object {
        "pcStat0": Object {
          "min": 0,
-         "max": 10,
+         "max": 1,
          "default": 3,
          "name": "PC Statisitic Zero",
        },"
`;

exports[`reducer > should set a stat minimum 1`] = `
"@@ -2,7 +2,7 @@
    "settings": Object {
      "pcStats": Object {
        "pcStat0": Object {
-         "min": 0,
+         "min": 1,
          "max": 10,
          "default": 3,
          "name": "PC Statisitic Zero","
`;

exports[`reducer > should set a stat name 1`] = `
"@@ -5,7 +5,7 @@
          "min": 0,
          "max": 10,
          "default": 3,
-         "name": "PC Statisitic Zero",
+         "name": "foo",
        },
      },
      "npcStats": Object {"
`;

exports[`reducer > should set a value using setSome 1`] = `
"@@ -61,7 +61,7 @@
        "world": Object {},
      },
      "useTurnPassingInitiative": false,
-     "genericOccupation": "",
+     "genericOccupation": "foo",
      "investigativeAbilityCategories": Array [],
      "longNotes": Array [],
      "mwHiddenShortNotes": Array [],"
`;
